nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,153 @@
+       * it from the One Big JSON metadata blob */
+      systemd.services.digitalocean-set-root-password = {
+        enable = cfg.setRootPassword;
+        path = [ pkgs.curl pkgs.shadow pkgs.jq ];
+        description = "Set root password provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          ROOT_PASSWORD=$(curl --retry-connrefused http://169.254.169.254/metadata/v1.json | jq -r '.auth_key')
+          echo "root:$ROOT_PASSWORD" | chpasswd
+          '';
+        unitConfig = {
+          After = [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Set the hostname from Digital Ocean, unless the user configured it in
+       * the NixOS configuration */
+      systemd.services.digitalocean-set-hostname = {
+        enable = hostName == "";
+        path = [ pkgs.curl pkgs.nettools ];
+        description = "Set hostname provided by Digitalocean";
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          set -e
+          DIGITALOCEAN_HOSTNAME=$(curl --retry-connrefused http://169.254.169.254/metadata/v1/hostname)
+          hostname $DIGITALOCEAN_HOSTNAME
+        '';
+        unitConfig = {
+          After =  [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+        serviceConfig = {
+          Type = "oneshot";
+        };
+      };
+
+      /* Fetch the ssh keys for root from Digital Ocean */
+      systemd.services.digitalocean-ssh-keys = {
+        enable = cfg.setSshKeys;
+        description = "Set root ssh keys provided by Digital Ocean";
+        wantedBy = [ "multi-user.target" ];
+        path = [ pkgs.curl ];
+        script = ''
+          set -e
+          mkdir -m 0700 -p /root/.ssh
+          curl --retry-connrefused --output /root/.ssh/authorized_keys http://169.254.169.254/metadata/v1/public-keys
+          chmod 600 /root/.ssh/authorized_keys
+        '';
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        };
+        unitConfig = {
+          ConditionFileExists = "!/root/.ssh/authorized_keys";
+          After =  [ "network-online.target" ];
+          Wants = [ "network-online.target" ];
+        };
+      };
+
+      /* Initialize the RNG by running the entropy-seed script from the
+       * Digital Ocean metadata
+       */
+      systemd.services.digitalocean-entropy-seed = {
+        enable = cfg.seedEntropy;
+        description = "Run the kernel RNG entropy seeding script from the Digital Ocean vendor data";
+        wantedBy = [ "multi-user.target" ];
+        path = [ pkgs.curl pkgs.mpack ];
+        script = ''
+          set -e
+          TEMPDIR=$(mktemp -d)
+          curl --retry-connrefused http://169.254.169.254/metadata/v1/vendor-data | munpack -C $TEMPDIR
+          $TEMPDIR/entropy-seed
```

Comment:
There is an error sometimes when this script runs, which causes issues in config switching when importing this module.

Reply:
is it the `curl` part that is failing or is it the `$TEMPDIR/entropy-seed` part that is failing?

Reply:
I think it is the curl not getting data, which is causing munpack to fail. But I'll have to try it again after work.

It didn't seem to be stopping the base system from coming up from the image, but if I build a system closure that imports digital-ocean-config.nix (I was trying this for my website) and I don't turn off the entropy seed fetcher, it gives an error when switching to it.

Reply:
Actually this is a known issue with munpack, since 2005 at least! https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=211472

That bug report from 2005 is talking about the same version of munpack packaged by nixpkgs now. I think we need an alternate solution, if we are going to grab the entropy data at all.

Reply:
For what it's worth, I'm using the following variant:

```bash
curl http://169.254.169.254/metadata/v1/vendor-data | munpack -tC $TEMPDIR
ENTROPY_SEED=$(grep -rl "DigitalOcean Entropy Seed script" $TEMPDIR)
${pkgs.runtimeShell} $ENTROPY_SEED
```